src/KOKKOS/fix_rigid_kokkos.cpp

```diff
@@ -0,0 +1,1274 @@
+    auto l_angmom = k_angmom.d_view;
+    auto l_omega = k_omega.d_view;
+    
+    auto l_fcm   = k_fcm.d_view;
+    auto l_fflag = k_fflag.d_view;
+    
+    auto l_torque = k_torque.d_view;
+    auto l_tflag  = k_tflag.d_view;
+    auto l_inertia = k_inertia.d_view;
+
+    auto l_ex_space = k_ex_space.d_view;
+    auto l_ey_space = k_ey_space.d_view;
+    auto l_ez_space = k_ez_space.d_view;
+
+    auto l_masstotal = masstotal;
+    
+    Kokkos::parallel_for(nbody, LAMMPS_LAMBDA(const int &ibody) {
+      // update vcm by 1/2 step
+
+      auto vcm_ibody = Kokkos::subview(l_vcm, ibody, Kokkos::ALL);
+      auto fcm_ibody = Kokkos::subview(l_fcm, ibody, Kokkos::ALL);
+      auto fflag_ibody = Kokkos::subview(l_fflag, ibody, Kokkos::ALL);
+      
+      auto angmom_ibody = Kokkos::subview(l_angmom, ibody, Kokkos::ALL);
+      auto omega_ibody = Kokkos::subview(l_omega, ibody, Kokkos::ALL);
+      
+      auto torque_ibody = Kokkos::subview(l_torque, ibody, Kokkos::ALL);
+      auto tflag_ibody = Kokkos::subview(l_tflag, ibody, Kokkos::ALL);
+      auto inertia_ibody = Kokkos::subview(l_inertia, ibody, Kokkos::ALL);
+
+      auto ex_space_ibody = Kokkos::subview(l_ex_space, ibody, Kokkos::ALL);
+      auto ey_space_ibody = Kokkos::subview(l_ey_space, ibody, Kokkos::ALL);
+      auto ez_space_ibody = Kokkos::subview(l_ez_space, ibody, Kokkos::ALL);
+      
+      double dtfm = dtf / l_masstotal[ibody];
+      vcm_ibody[0] += dtfm * fcm_ibody[0] * fflag_ibody[0];
+      vcm_ibody[1] += dtfm * fcm_ibody[1] * fflag_ibody[1];
+      vcm_ibody[2] += dtfm * fcm_ibody[2] * fflag_ibody[2];
+      
+      // update angular momentum by 1/2 step
+      
+      angmom_ibody[0] += dtf * torque_ibody[0] * tflag_ibody[0];
+      angmom_ibody[1] += dtf * torque_ibody[1] * tflag_ibody[1];
+      angmom_ibody[2] += dtf * torque_ibody[2] * tflag_ibody[2];
+      
+      MathExtraKokkos::angmom_to_omega<DeviceType>(l_angmom,l_ex_space,
+                                                   l_ey_space, l_ez_space,
+                                                   l_inertia,l_omega, ibody);
+
+    });
+  }
+  
+
+  // set velocity/rotation of atoms in rigid bodies
+  // virial is already setup from initial_integrate
+  if(debug_output && comm->me == 0) {
+    fprintf(stderr, "After parallel_for, before set_v, we have:\n");
+  }
+  
+  debug_print_vec(k_vcm, vcm, "vcm");
+  debug_print_vec(k_omega, omega, "omega");
+  
+  set_v_kokkos();
+  atomKK->sync(execution_space, datamask_read);
+}
+
+
+
+template <class DeviceType>
+void FixRigidKokkos<DeviceType>::compute_forces_and_torques_kokkos()
+{
+  // TODO: PORT THIS; CONTINUE HERE LATER.
+  int i,ibody;
+  
+  // sum over atoms to get force and torque on rigid body
+
+  double **x = atom->x;
+  double **f = atom->f;
+  int nlocal = atom->nlocal;
+
+  double dx,dy,dz;
+  double unwrap[3];
+
+  for (ibody = 0; ibody < nbody; ibody++)
+    for (i = 0; i < 6; i++) sum[ibody][i] = 0.0;
+
+  for (i = 0; i < nlocal; i++) {
```

Comment:
Will need to use a `Kokkos:: parallel_reduce`

Reply:
I don't know how to properly use that. I saw in various places code where Kokkos::parallel_reduce is used to reduce arrays to a single variable (e.g, a full sum), but in this case, the sum goes over all atoms on all processors that might be in the same body, and then they are all reduced to an array that is still over all bodies.
I wrote it in terms of a Kokkos::parallel_for and then an MPI_Allreduce, but I do not know if that is the right way to do it. I can imagine the mandatory syncing to host before the MPI_Allreduce is not ideal.

Reply:
Actually I looked at this more closely and the `sum` view needs to be atomic (more precisely a `Kokkos::ScatterView`, but I can add that later). Then you can pass the device `d_view.data()` pointers directly in the `MPI_Allreduce` call if the `cuda_aware_mpi` flag is enabled. Otherwise you will need to do the sync to host and then pass the host pointer to the Allreduce.

Reply:
For example, see https://github.com/lammps/lammps/blob/master/src/KOKKOS/gridcomm_kokkos.cpp#L532